src/gen_bpf.c

```diff
@@ -1222,111 +1229,257 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
 }
 
 /**
- * Generate the BPF instruction blocks for a given filter/architecture
- * @param state the BPF state
- * @param db the filter DB
- * @param db_secondary the secondary DB
+ * Calculate the number of levels in the binary tree
+ * @param syscall_cnt the number of syscalls in this seccomp filter
+ */
+static int _get_bintree_levels(unsigned int syscall_cnt,
+			       uint32_t optimize)
+{
+	unsigned int i = 2, max_level = SYSCALLS_PER_NODE * 2;
+
+	if (syscall_cnt < MIN_SYSCALLS_TO_USE_BINTREE ||
+	    optimize != 2)
+		/* Only use a binary tree if there are a lot of syscalls
+		 * and if the filter optimization level is set to 2 (aka
+		 * binary tree)
+		 */
+		return 0;
+
+	while(max_level < syscall_cnt) {
+		max_level <<= 1;
+		i++;
+	}
+
+	return i;
+}
+
+/**
+ * Sort the syscalls by priority
+ * @param syscalls the linked list of syscalls to be sorted
+ * @param s_head the head of the linked list to be returned to the caller
+ * @param s_tail the tail of the linked list to be returned to the caller
  *
- * Generate the BPF instruction block for the given filter DB(s)/architecture(s)
- * and return a pointer to the block on succes, NULL on failure.  The resulting
- * block assumes that the architecture token has already been loaded into the
- * BPF accumulator.
+ * Sorts syscalls by their syscall number
  *
  */
-static struct bpf_blk *_gen_bpf_arch(struct bpf_state *state,
-				     const struct db_filter *db,
-				     const struct db_filter *db_secondary)
+static void _sort_syscalls_by_priority(struct db_sys_list *syscalls,
```

Comment:
This is some good quality bikeshedding here, but I'm not typically a fan of function-names-as-sentences, how about `_sys_priority_sort(...)`?

Reply:
Works for me.  Sometimes I fall into the Java trap of overly describing everything ;)